server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -0,0 +1,686 @@
+            // keeping them in the same file indicates we pull the entire file into cache
+            docIdsWriter.writeDocIds(j -> floatVectorValues.ordToDoc(cluster.get(j)), size, postingsOutput);
+            writePostingList(cluster, postingsOutput, binarizedByteVectorValues);
+        }
+        return offsets;
+    }
+
+    private static void printClusterQualityStatistics(IntArrayList[] clusters, InfoStream infoStream) {
+        float min = Float.MAX_VALUE;
+        float max = Float.MIN_VALUE;
+        float mean = 0;
+        float m2 = 0;
+        // iteratively compute the variance & mean
+        int count = 0;
+        for (IntArrayList cluster : clusters) {
+            count += 1;
+            if (cluster == null) {
+                continue;
+            }
+            float delta = cluster.size() - mean;
+            mean += delta / count;
+            m2 += delta * (cluster.size() - mean);
+            min = Math.min(min, cluster.size());
+            max = Math.max(max, cluster.size());
+        }
+        float variance = m2 / (clusters.length - 1);
+        infoStream.message(
+            IVF_VECTOR_COMPONENT,
+            "Centroid count: "
+                + clusters.length
+                + " min: "
+                + min
+                + " max: "
+                + max
+                + " mean: "
+                + mean
+                + " stdDev: "
+                + Math.sqrt(variance)
+                + " variance: "
+                + variance
+        );
+    }
+
+    static void assignCentroids(CentroidAssignmentScorer scorer, FloatVectorValues vectors, IntArrayList[] clusters) throws IOException {
+        int numCentroids = scorer.size();
+        // we at most will look at the EXT_SOAR_LIMIT_CHECK_RATIO nearest centroids if possible
+        int soarToCheck = (int) (numCentroids * EXT_SOAR_LIMIT_CHECK_RATIO);
+        int soarClusterCheckCount = Math.min(numCentroids - 1, soarToCheck);
+        NeighborQueue neighborsToCheck = new NeighborQueue(soarClusterCheckCount + 1, true);
+        OrdScoreIterator ordScoreIterator = new OrdScoreIterator(soarClusterCheckCount + 1);
+        float[] scratch = new float[vectors.dimension()];
+        for (int docID = 0; docID < vectors.size(); docID++) {
+            float[] vector = vectors.vectorValue(docID);
+            scorer.setScoringVector(vector);
+            int bestCentroid = 0;
+            float bestScore = Float.MAX_VALUE;
+            if (numCentroids > 1) {
+                for (short c = 0; c < numCentroids; c++) {
+                    float squareDist = scorer.score(c);
+                    neighborsToCheck.insertWithOverflow(c, squareDist);
+                }
+                // pop the best
+                int sz = neighborsToCheck.size();
+                int best = neighborsToCheck.consumeNodesAndScoresMin(ordScoreIterator.ords, ordScoreIterator.scores);
+                // reset the ordScoreIterator as it has consumed the ords and scores
```

Comment:
Why is reset needed here?

Reply:
This indicates to the `ordScoreIterator` that it contains these values as we are accessing the underlying arrays directly. Let me rename this. I realize the logic here isn't clear.